### PR TITLE
714 push translations

### PR DESCRIPTION
**Pull request by clemens-msupply**
I also added some sync docs. I used the term messages for incoming remote data from other sites. Happy to change it if there is a better term.

### CODE REVIEW

File: server/src/sync/translation_remote/invoice.rs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+            linked_transaction_id: linked_invoice_id,
+            entry_date: date_from_date_time(&created_datetime),
+            entry_time: time_sec_from_date_time(&created_datetime),
+            // TODO losing the time here:
+            ship_date: shipped_datetime
+                .map(|shipped_datetime| date_from_date_time(&shipped_datetime)),
+            // TODO losing the time here:
+            arrival_date_actual: delivered_datetime
+                .map(|delivered_datetime| date_from_date_time(&delivered_datetime)),
+            confirm_date: verified_datetime
```
Comment (andreievg) [inline server/src/sync/translation_remote/invoice.rs:310]: confirm is different based on the type

File: server/src/sync/translation_remote/name_store_join.rs [hunk 2]
```diff
@@ -69,3 +70,46 @@ impl RemotePullTranslation for NameStoreJoinTranslation {
...
+        &self,
+        connection: &StorageConnection,
+        changelog: &ChangelogRow,
+    ) -> Result<Option<Vec<PushUpsertRecord>>, SyncTranslationError> {
+        if changelog.table_name != ChangelogTableName::NameStoreJoin {
+            return Ok(None);
+        }
+        let table_name = TRANSLATION_RECORD_NAME_STORE_JOIN;
+
+        let NameStoreJoinRow {
```
Comment (andreievg) [inline server/src/sync/translation_remote/name_store_join.rs:85]: not pushable for normal names (might something else for patients).

File: server/src/sync/translation_remote/push.rs [hunk 3]
```diff
@@ -58,8 +63,15 @@ pub fn translate_changelog(
...
-            let translations: Vec<Box<dyn RemotePushUpsertTranslation>> =
-                vec![Box::new(NumberTranslation {})];
+            let translations: Vec<Box<dyn RemotePushUpsertTranslation>> = vec![
+                Box::new(NumberTranslation {}),
+                Box::new(StockLineTranslation {}),
+                Box::new(NameStoreJoinTranslation {}),
+                Box::new(InvoiceTranslation {}),
+                Box::new(InvoiceLineTranslation {}),
+                Box::new(StocktakeTranslation {}),
+                Box::new(StocktakeLineTranslation {}),
```
Comment (andreievg) [inline server/src/sync/translation_remote/push.rs:73]: Nice

File: server/src/sync/translation_remote/stocktake_line.rs [hunk 4]
```diff
@@ -86,3 +87,91 @@ impl RemotePullTranslation for StocktakeLineTranslation {
...
+            stock_take_qty: counted_number_of_packs.unwrap_or(0),
+            is_edited: counted_number_of_packs.is_some(),
+            item_line_ID: stock_line_id,
+            item_ID: item_id,
+            snapshot_packsize: stock_line
+                .as_ref()
+                .map(|it| it.pack_size)
+                .or(pack_size)
+                .unwrap_or(0),
+            Batch: stock_line
```
Comment (andreievg) [inline server/src/sync/translation_remote/stocktake_line.rs:149]: I think this should just be batch ?

File: server/src/sync/translation_remote/stocktake_line.rs [hunk 5]
```diff
@@ -86,3 +87,91 @@ impl RemotePullTranslation for StocktakeLineTranslation {
...
+        } = StocktakeLineRowRepository::new(connection)
+            .find_one_by_id(&changelog.row_id)
+            .map_err(|err| to_push_translation_error(table_name, err.into(), changelog))?
+            .ok_or(to_push_translation_error(
+                table_name,
+                anyhow::Error::msg("Stocktake row not found"),
+                changelog,
+            ))?;
+
+        let stock_line = match &stock_line_id {
```
Comment (andreievg) [inline server/src/sync/translation_remote/stocktake_line.rs:126]: I think this is not needed ?

**Review by andreievg — APPROVED** [review]
Name store joins are not pushable i think.

I want to do a full review of translations once we've added extra fields in mSupply, btw it's easy to do, I can show you

File: server/src/sync/translation_remote/name_store_join.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/src/sync/translation_remote/name_store_join.rs:85]: Ok so I remove it completely for now, right?

File: server/src/sync/translation_remote/stocktake_line.rs (hunk 5 above)
Comment (clemens-msupply) [reply server/src/sync/translation_remote/stocktake_line.rs:126]: I wasn't sure what mSupply expects but yes it's probably not needed. Not sure about snapshot_packsize..

File: server/src/sync/translation_remote/invoice.rs (hunk 1 above)
Comment (clemens-msupply) [reply server/src/sync/translation_remote/invoice.rs]: thanks I reworked the mapping (for pull and push) Still not 100% sure about it. Would also need some more tests once finalized.

File: server/src/sync/translation_remote/name_store_join.rs (hunk 2 above)
Comment (andreievg) [reply server/src/sync/translation_remote/name_store_join.rs:85]: yep

**Pull request merged**